internal/controller/service_controller.go

```diff
@@ -0,0 +1,90 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	annotations = ".metadata.annotations"
+)
+
+// ServiceReconciler reconciles a Service object
+type ServiceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme // 리소스의 타입 정보를 관리
+}
+
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=services/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=services/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Service object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/reconcile
+func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	instance := &corev1.Service{}
+	if instance != nil {
+		log.Info("service exists", "annotated tunnel name", instance.Annotations["tunnel.name"])
+	}
+	log.Info("Reconciling Service", "service", req.NamespacedName)
+	// TODO(user): your logic here
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Service{}).
+		WithEventFilter(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldAnnotations := e.ObjectOld.GetAnnotations()
+				newAnnotations := e.ObjectNew.GetAnnotations()
+				// if no changes in annotations, return false
+				if reflect.DeepEqual(oldAnnotations, newAnnotations) {
+					return false
+				}
+				if tunnelName, exists := newAnnotations["tunnel.name"]; exists {
+					if len(oldAnnotations["tunnel.name"]) == 0 {
+						return true
+					} else {
+						if tunnelName != oldAnnotations["tunnel.name"] {
+							// TODO: Decide what to do if tunnel.name already exists but updated with diff value
+						}
+					}
+				}
+				return false
```

Comment:
```suggestion
				tunnelName, exists := newAnnotations["tunnel.name"]
				if !exists {
					return false
				}
				if len(oldAnnotations["tunnel.name"]) == 0 {
					return true
				}
				if tunnelName != oldAnnotations["tunnel.name"] {
					// TODO: Decide what to do if tunnel.name already exists but updated with diff value
				}
				return false
```

분기문이 깊어지면 가독성이 떨어져서 early return하는게 어떨까?

Reply:
넵 수정하겠습니다!